File: Algorithm/leetcode_048.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
# File              : leetcode_048.py
# Author            : Andy
# Date              : 2022.08.31
# Last Modified Date: 2022.08.31
# Last Modified By  : Andy


class Solution:

    def rotate(self, matrix):
        rows, cols = len(matrix), len(matrix[0])

        def transpost(matrix):
            # 行列转置
            # TODO j为什么是到i+1,目前还没太明白
            for i in range(rows):
                for j in range(i + 1, cols):
                    matrix[i][j], matrix[j][i] = matrix[j][i], matrix[i][j]

        def reverse(matrix):
            # 前后倒置
            for i in range(len(matrix)):
                matrix[i] = matrix[i][::-1]

        transpost(matrix)
        reverse(matrix)

File: Algorithm/leetcode_048.go
/**
 * File              : leetcode_048.go
 * Author            : Andy
 * Date              : 2022.08.31
 * Last Modified Date: 2022.08.31
 * Last Modified By  : Andy
 */

package rotate

func transpost(matrix *[][]int) {
	rows := len(*matrix)
	cols := len((*matrix)[0])
	for i := 0; i < rows; i++ {
		for j := i + 1; j < cols; j++ {
			item := (*matrix)
			item[i][j], item[j][i] = item[j][i], item[i][j]
		}
	}
}

func reverse(matrix, *[][]int) {
	rows := len((*matrix))
	for i := 0; i < rows; i++ {
		tmp := (*matrix)[i]
		// 倒置有点麻烦
		for m, n := 0, len(tmp)-1; m < n; m, n = m+1, n-1 {
			tmp[m], tmp[n] = tmp[n], tmp[m]
		}
	}
}

func rotate(matrix [][]int) {
	transpost(&matrix)
	reverse(&matrix)
}
